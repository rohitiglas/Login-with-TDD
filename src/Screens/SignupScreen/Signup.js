import React, {useState} from 'react';
import {Alert, Button, TextInput, Text, View, StyleSheet,TouchableOpacity,StatusBar} from 'react-native';
import * as regex from '../../Regex/regex'
import styles from "./styles";

export const ValidationErrors = {
    FormEmpty: 'Form fields cannot be blank',
    NameEmpty: 'Name cannot be blank',
    EmailEmpty: 'Email cannot be blank',
    MobileNumberEmpty: 'Mobile number cannot be blank',
    UsernameEmpty: 'Username cannot be blank',
    PasswordEmpty: 'Password cannot be blank',
    ConfirmPasswordEmpty: 'Confirm Password cannot be blank',
    PasswordMatch: 'Password should be match',
    ValidEmail: 'Please enter valid Email',
    ValidMobile: 'Please enter valid mobile',
    ValidPassword: 'Password should contain min 8 char with one special, one small , one Capital & one Number.',
    PasswordShouldMatch:'Password does not match'

};
const Signup = () => {
    const [name, setName] = useState('');
    const [email, setEmail] = useState('');
    const [mobileNumber, setMobileNumber] = useState('');
    const [userName, setUserName] = useState('');
    const [password, setPassword] = useState('');
    const [confirmPassword, setConfirmPassword] = useState('');
    const [validationError, setValidationError] = useState('');
    const [validateRegex, setValidateRegex]= useState({})

    const onLogin = () => {
        if (isFormEmpty()) {
            setValidationError(ValidationErrors.FormEmpty);
        } else if (name.length === 0) {
            setValidationError(ValidationErrors.NameEmpty);
        } else if (email.length === 0) {
            setValidationError(ValidationErrors.EmailEmpty);
        } else if (mobileNumber.length === 0) {
            setValidationError(ValidationErrors.MobileNumberEmpty);
        } else if (userName.length === 0) {
            setValidationError(ValidationErrors.UsernameEmpty);
        } else if (password.length === 0) {
            setValidationError(ValidationErrors.PasswordEmpty);
        } else if (confirmPassword.length === 0) {
            setValidationError(ValidationErrors.ConfirmPasswordEmpty);
        }
        else {
            Alert.alert('Credentials', `${userName} + ${password}`);
        }
    };

    const isFormEmpty=()=>{
        return (name.length===0 && email.length===0 &&
            mobileNumber.length===0 &&
            userName.length===0 && password.length===0);
    }

   const regexValidation=(type, value)=>{
    if(type==='input-email')
    {
        setEmail(value);
        setValidateRegex({...validateRegex, [type]: regex.emailRegex.test(value)})
    } 
    if(type==='input-mobile'){
        setMobileNumber(value)
        setValidateRegex({...validateRegex, [type]: regex.mobileRegex.test(value)})
    } 
    if(type==='input-password'){
        setPassword(value)
        setValidateRegex({...validateRegex, [type]: regex.passwordRegex.test(value)})
    } 
    if(type==='input-confirm-password'){
        setConfirmPassword(value)
        setValidateRegex({...validateRegex, [type]: regex.passwordRegex.test(value)})
    } 
   }

    return (
        <View style={styles.mainContainer}>
            {validationError.length !== 0 && (
                <Text testID={'text-error'} style={styles.errorText}>
                    {validationError}
                </Text>
            )}

            <View style={styles.loginContainer}>
                <StatusBar barStyle="light-content"/>
                <TextInput
                    testID={'input-name'}
                    value={name}
                    onChangeText={(text) => setName(text)}
                    placeholder={'Name'}
                    style={styles.input}
                />
             

                <TextInput
                    testID={'input-email'}
                    value={email}
                    onChangeText={(text) => regexValidation('input-email',text)}
                    placeholder={'Email Id'}
                    style={styles.input}
                />
                {validateRegex && !validateRegex['input-email']&& email.length>0 &&
                  <Text
                      testID={'input-email-error'}
                      style={{top:-20,  alignItems:'flex-start', color:'red'}}>
                      {ValidationErrors.ValidEmail}
                </Text>
                }
              
                <TextInput
                    testID={'input-mobile'}
                    maxLength={10}
                    keyboardType = 'numeric'
                    value={mobileNumber}
                    onChangeText={(text) => regexValidation('input-mobile', text)}
                    placeholder={'Mobile Number'}
                    style={styles.input}
                />
                {validateRegex && !validateRegex['input-mobile']&& mobileNumber.length >0 &&
                 <Text style={{top:-20,  alignItems:'flex-start', color:'red'}}>
                     {ValidationErrors.ValidMobile}
               </Text>}
                <TextInput
                    testID={'input-username'}
                    value={userName}
                    onChangeText={(text) => setUserName(text)}
                    placeholder={'UserName'}
                    style={styles.input}
                />
                <TextInput
                    testID={'input-password'}
                    value={password}
                    onChangeText={(text) => regexValidation('input-password', text)}
                    placeholder={'Password'}
                    secureTextEntry={true}
                    style={styles.input}
                />
                {validateRegex && !validateRegex['input-password']&& password.length >0 &&
                 <Text style={{top:-20,  alignItems:'flex-start', color:'red'}}>
                     {ValidationErrors.ValidPassword}
               </Text>
                }
                <TextInput
                    testID={'input-confirm-password'}
                    value={confirmPassword}
                    onChangeText={(text) => regexValidation('input-confirm-password', text)}
                    placeholder={'Confirm Password'}
                    secureTextEntry={true}
                    style={styles.input}
                />
                {validateRegex && !validateRegex['input-mobile']&& confirmPassword.length >0 && password === confirmPassword &&
               <Text style={{top:-20,  alignItems:'flex-start', color:'red'}}>
                   {ValidationErrors.PasswordShouldMatch}
               </Text>}
                <TouchableOpacity
                    testID={'submit-button'}
                    style={styles.buttonContainer}
                    onPress={onLogin}>
                    <Text  style={styles.buttonText}>Register</Text>
                </TouchableOpacity>
            </View>
        </View>

    );
};

export default Signup;