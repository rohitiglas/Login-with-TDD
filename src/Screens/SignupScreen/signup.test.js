import 'react-native';
import React from 'react';
import Signup,{ValidationErrors} from "./Signup";
import {render,cleanup,fireEvent} from "@testing-library/react-native";
import {Alert} from "react-native";

jest.mock('react-native/Libraries/Alert/Alert', () => ({
    alert: jest.fn(),
}));

describe('Signup Form',()=>{
    let wrapper;
    beforeEach(()=>{
      wrapper=render(<Signup/>);
    });

    afterEach(()=>{
       cleanup();
       wrapper=null;
    });


     //UI Testing
     it('Should display renders correctly',()=>{
         expect(wrapper).toMatchSnapshot();
     });

    it("shows 'form-fields-empty' error if All field empty", () => {
        const {getByTestId} = wrapper;
        const submitButton = getByTestId('submit-button');
        fireEvent.press(submitButton);
        const validationError = getByTestId('text-error');
        expect(validationError).toBeTruthy();
        expect(validationError.props.children)
            .toBe(ValidationErrors.FormEmpty);
    });

    it('Show error if user left Name field',()=>{
        const {getByTestId}=wrapper;
        const emailText=getByTestId('input-email');
        const email='rohitiglas2@gmail.com';
        fireEvent(emailText,'onChangeText',email);
        const mobileText=getByTestId('input-mobile');
        const mobile='9643198239';
        fireEvent(mobileText,'onChangeText',mobile);
        const userNameText=getByTestId('input-username');
        const userName='RohitBansal';
        fireEvent(userNameText,'onChangeText',userName);
        const passwordText=getByTestId('input-password');
        const password='123456';
        fireEvent(passwordText,'onChangeText',password);
        const confirmPasswordText=getByTestId('input-confirm-password');
        const confirmPassword='123456';
        fireEvent(confirmPasswordText,'onChangeText',confirmPassword);
        const submitButton=getByTestId('submit-button');
        fireEvent(submitButton, 'onPress');
        const validationError = getByTestId('text-error');
        expect(validationError).toBeTruthy();
        expect(validationError.props.children).toBe(ValidationErrors.NameEmpty);
    });

    it('Show error if user left email field',()=>{
        const {getByTestId}=wrapper;
        const nameText=getByTestId('input-name');
        const name='Rohit Bansal';
        fireEvent(nameText,'onChangeText',name);
        const mobileText=getByTestId('input-mobile');
        const mobile='9643198239';
        fireEvent(mobileText,'onChangeText',mobile);
        const userNameText=getByTestId('input-username');
        const userName='RohitBansal';
        fireEvent(userNameText,'onChangeText',userName);
        const passwordText=getByTestId('input-password');
        const password='123456';
        fireEvent(passwordText,'onChangeText',password);
        const confirmPasswordText=getByTestId('input-confirm-password');
        const confirmPassword='123456';
        fireEvent(confirmPasswordText,'onChangeText',confirmPassword);
        const submitButton=getByTestId('submit-button');
        fireEvent(submitButton, 'onPress');
        const validationError = getByTestId('text-error');
        expect(validationError).toBeTruthy();
        expect(validationError.props.children).toBe(ValidationErrors.EmailEmpty);
    });

    it('Show error if user left mobile field',()=>{
        const {getByTestId}=wrapper;
        const emailText=getByTestId('input-email');
        const email='rohitiglas2@gmail.com';
        fireEvent(emailText,'onChangeText',email);
        const nameText=getByTestId('input-name');
        const name='Rohit Bansal';
        fireEvent(nameText,'onChangeText',name);
        const userNameText=getByTestId('input-username');
        const userName='RohitBansal';
        fireEvent(userNameText,'onChangeText',userName);
        const passwordText=getByTestId('input-password');
        const password='123456';
        fireEvent(passwordText,'onChangeText',password);
        const confirmPasswordText=getByTestId('input-confirm-password');
        const confirmPassword='123456';
        fireEvent(confirmPasswordText,'onChangeText',confirmPassword);
        const submitButton=getByTestId('submit-button');
        fireEvent(submitButton, 'onPress');
        const validationError = getByTestId('text-error');
        expect(validationError).toBeTruthy();
        expect(validationError.props.children).toBe(ValidationErrors.MobileNumberEmpty);
    });

    it('Show error if user left userName field',()=>{
        const {getByTestId}=wrapper;
        const emailText=getByTestId('input-email');
        const email='rohitiglas2@gmail.com';
        fireEvent(emailText,'onChangeText',email);
        const nameText=getByTestId('input-name');
        const name='Rohit Bansal';
        fireEvent(nameText,'onChangeText',name);
        const mobileText=getByTestId('input-mobile');
        const mobile='9643198239';
        fireEvent(mobileText,'onChangeText',mobile);
        const passwordText=getByTestId('input-password');
        const password='123456';
        fireEvent(passwordText,'onChangeText',password);
        const confirmPasswordText=getByTestId('input-confirm-password');
        const confirmPassword='123456';
        fireEvent(confirmPasswordText,'onChangeText',confirmPassword);
        const submitButton=getByTestId('submit-button');
        fireEvent(submitButton, 'onPress');
        const validationError = getByTestId('text-error');
        expect(validationError).toBeTruthy();
        expect(validationError.props.children).toBe(ValidationErrors.UsernameEmpty);
    });

    it('Show error if user left password field',()=>{
        const {getByTestId}=wrapper;
        const emailText=getByTestId('input-email');
        const email='rohitiglas2@gmail.com';
        fireEvent(emailText,'onChangeText',email);
        const nameText=getByTestId('input-name');
        const name='Rohit Bansal';
        fireEvent(nameText,'onChangeText',name);
        const mobileText=getByTestId('input-mobile');
        const mobile='9643198239';
        fireEvent(mobileText,'onChangeText',mobile);
        const userNameText=getByTestId('input-username');
        const userName='RohitBansal';
        fireEvent(userNameText,'onChangeText',userName);
        const confirmPasswordText=getByTestId('input-confirm-password');
        const confirmPassword='123456';
        fireEvent(confirmPasswordText,'onChangeText',confirmPassword);
        const submitButton=getByTestId('submit-button');
        fireEvent(submitButton, 'onPress');
        const validationError = getByTestId('text-error');
        expect(validationError).toBeTruthy();
        expect(validationError.props.children).toBe(ValidationErrors.PasswordEmpty);
    });

    it('Show error if user left password field',()=>{
        const {getByTestId}=wrapper;
        const emailText=getByTestId('input-email');
        const email='rohitiglas2@gmail.com';
        fireEvent(emailText,'onChangeText',email);
        const nameText=getByTestId('input-name');
        const name='Rohit Bansal';
        fireEvent(nameText,'onChangeText',name);
        const mobileText=getByTestId('input-mobile');
        const mobile='9643198239';
        fireEvent(mobileText,'onChangeText',mobile);
        const userNameText=getByTestId('input-username');
        const userName='RohitBansal';
        fireEvent(userNameText,'onChangeText',userName);
        const passwordText=getByTestId('input-password');
        const password='123456';
        fireEvent(passwordText,'onChangeText',password);
        const submitButton=getByTestId('submit-button');
        fireEvent(submitButton, 'onPress');
        const validationError = getByTestId('text-error');
        expect(validationError).toBeTruthy();
        expect(validationError.props.children).toBe(ValidationErrors.ConfirmPasswordEmpty);
    });

    it('should show Alert dialog if form validation successful', async () => {
        const {getByTestId}=wrapper;
        const emailText=getByTestId('input-email');
        const email='rohitiglas2@gmail.com';
        fireEvent(emailText,'onChangeText',email);
        const nameText=getByTestId('input-name');
        const name='Rohit Bansal';
        fireEvent(nameText,'onChangeText',name);
        const mobileText=getByTestId('input-mobile');
        const mobile='9643198239';
        fireEvent(mobileText,'onChangeText',mobile);
        const userNameText=getByTestId('input-username');
        const userName='RohitBansal';
        fireEvent(userNameText,'onChangeText',userName);
        const passwordText=getByTestId('input-password');
        const password='123456';
        fireEvent(passwordText,'onChangeText',password);
        const confirmPasswordText=getByTestId('input-confirm-password');
        const confirmPassword='123456';
        fireEvent(confirmPasswordText,'onChangeText',confirmPassword);

        // Press Submit Button
        const submitButton = getByTestId('submit-button');
        fireEvent.press(submitButton);

        // Check if Alert.alert() has been called
        const alertSpy = jest.spyOn(Alert, 'alert');
        expect(alertSpy).toHaveBeenCalled();
    });
})