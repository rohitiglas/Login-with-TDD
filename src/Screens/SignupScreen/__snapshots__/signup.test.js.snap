// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Form Should display renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2c3e50",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexGrow": 1,
        "justifyContent": "center",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Name"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 20,
          "color": "#2c3e50",
          "height": 50,
          "marginBottom": 20,
          "padding": 10,
          "width": "80%",
        }
      }
      testID="input-name"
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Email Id"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 20,
          "color": "#2c3e50",
          "height": 50,
          "marginBottom": 20,
          "padding": 10,
          "width": "80%",
        }
      }
      testID="input-email"
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      keyboardType="numeric"
      maxLength={10}
      onChangeText={[Function]}
      placeholder="Mobile Number"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 20,
          "color": "#2c3e50",
          "height": 50,
          "marginBottom": 20,
          "padding": 10,
          "width": "80%",
        }
      }
      testID="input-mobile"
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="UserName"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 20,
          "color": "#2c3e50",
          "height": 50,
          "marginBottom": 20,
          "padding": 10,
          "width": "80%",
        }
      }
      testID="input-username"
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 20,
          "color": "#2c3e50",
          "height": 50,
          "marginBottom": 20,
          "padding": 10,
          "width": "80%",
        }
      }
      testID="input-password"
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Confirm Password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 20,
          "color": "#2c3e50",
          "height": 50,
          "marginBottom": 20,
          "padding": 10,
          "width": "80%",
        }
      }
      testID="input-confirm-password"
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#2980b6",
          "borderRadius": 20,
          "opacity": 1,
          "paddingVertical": 15,
          "width": "80%",
        }
      }
      testID="submit-button"
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontWeight": "700",
            "textAlign": "center",
          }
        }
      >
        Register
      </Text>
    </View>
  </View>
</View>
`;
